### PR TITLE
support topic partitions in kafka change feeds

### PR DESCRIPTION
**Pull request by snopoke**
Still outstanding:

* [ ] migrate all existing kafka checkpoints to new format

## Part II
* allow dynamic pillow configurations via `pillows/{env}.yml
* add more partitions to high volume topics

### CODE REVIEW

File: corehq/apps/change_feed/consumer/feed.py [hunk 1]
```diff
@@ -58,41 +59,27 @@ def iter_changes(self, since, forever):
...
-                    offset = 0
-                self._processed_topic_offsets = {single_topic: offset}
-
-            def _make_offset_tuple(topic):
-                if topic in self._processed_topic_offsets:
-                    return (topic, self._partition, self._processed_topic_offsets[topic])
-                else:
-                    return (topic, self._partition)
-
-            offsets = [_make_offset_tuple(topic) for topic in self._topics]
```
Comment (benrudolph) [inline corehq/apps/change_feed/consumer/feed.py]: nice cleanup

**Review by benrudolph — COMMENTED** [review]
cool, looking good. i had a bit of a hard time keeping all the changes in my head, but in general looks like the correct approach

**Review by emord — COMMENTED** [review]
Looking good. I really like the change to all change feeds using multiple topics instead of having the single and multiple topics

When adding partitions are we able to do it on the fly without downtime, or will it require more configuration?

**Review by mkangia — APPROVED** [review]

**Comment by snopoke** [comment]
I think this is ready to be merged

File: corehq/apps/cleanup/pillow_migrations.py [hunk 2]
```diff
@@ -91,3 +91,64 @@ def _merge_seq(topic, seq):
...
+
+def migrate_kafka_checkpoints(apps, schema_editor):
+    from pillowtop.utils import get_kafka_pillow_instances
+    DjangoPillowCheckpoint = apps.get_model('pillowtop', 'DjangoPillowCheckpoint')
+
+    for pillow in get_kafka_pillow_instances():
+        try:
+            checkpoint = DjangoPillowCheckpoint.objects.get(checkpoint_id=pillow.checkpoint.checkpoint_id)
+            print("Migrating checkpoint {}: {} - {}".format(
+                checkpoint.checkpoint_id, change_feed.topics, checkpoint.sequence
```
Comment (dimagimon) [inline corehq/apps/cleanup/pillow_migrations.py]: F821 undefined name 'change_feed'

File: corehq/apps/cleanup/tests/test_migrate_kafka_checkpoint_formats.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    pass
+
+
+@generate_cases([
+    ('text', FakeFeed(['t1']), "123", {('t1', 0): 123}),
+    ('json', FakeFeed(['t1']), "123", {('t1', 0): 123}),
+    ('text', FakeFeed(['t1', 't2']), "123", None, AssertionError),
+    ('json', FakeFeed(['t1', 't2']), "123", None, AssertionError),
+    ('json', FakeFeed(['t1', 't2']), "abc", None, ValueError),
+    ('json', FakeFeed(['t1', 't2']), '{"t1": 123, "t2": 345}', {('t1', 0): 123, ('t2', 0): 345}),
```
Comment (benrudolph) [inline corehq/apps/cleanup/tests/test_migrate_kafka_checkpoint_formats.py]: 👍

**Review by benrudolph — APPROVED** [review]

File: corehq/apps/cleanup/pillow_migrations.py [hunk 4]
```diff
@@ -91,3 +91,64 @@ def _merge_seq(topic, seq):
...
+    try:
+        # optimistically try convert to int
+        int_seq = int(checkpoint.sequence)
+    except ValueError:
+        pass
+
+    if checkpoint.sequence_format == 'text' or int_seq is not None:
+        topics = change_feed.topics
+        assert len(topics) == 1, topics
+        return {(topics[0], 0): int_seq}
```
Comment (benrudolph) [inline corehq/apps/cleanup/pillow_migrations.py]: should these use `TopicAndPartition`?

Comment (benrudolph) [inline corehq/apps/cleanup/pillow_migrations.py]: ah nvm this is getting set directly on the model

File: corehq/pillows/domain.py [hunk 5]
```diff
@@ -37,13 +37,14 @@ def get_domain_kafka_to_elasticsearch_pillow(pillow_id='KafkaDomainPillow'):
         index_info=DOMAIN_INDEX_INFO,
         doc_prep_fn=transform_domain_for_elasticsearch
     )
+    change_feed = KafkaChangeFeed(topics=[DOMAIN], group_id='domains-to-es')
```
Comment (emord) [inline corehq/pillows/domain.py]: pillows for domain and group are getting them from document_types file instead of topics.

**Comment by emord** [comment]
failure looks related (cut irrelevant parts)

```
ERROR: Failure: OperationalError (FATAL:  database "test_commcarehq" does not exist
)
...
  File "/mnt/commcare-hq-ro/corehq/apps/commtrack/tests/test_xml.py", line 212, in <module>
    class CommTrackSubmissionTest(XMLTest):
  File "/mnt/commcare-hq-ro/corehq/apps/commtrack/tests/test_xml.py", line 220, in CommTrackSubmissionTest
    @process_pillow_changes('LedgerToElasticsearchPillow')
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/utils.py", line 17, in __init__
    self.pillow = get_pillow_by_name(pillow_name_or_instance, instantiate=True)
  File "corehq/ex-submodules/pillowtop/utils.py", line 109, in get_pillow_by_name
    return config.get_instance() if instantiate else config.get_class()
  File "corehq/ex-submodules/pillowtop/utils.py", line 80, in get_instance
    return instance_generator_fn(self.name, **self.params)
  File "/mnt/commcare-hq-ro/corehq/pillows/ledger.py", line 73, in get_ledger_to_elasticsearch_pillow
    checkpoint=checkpoint, checkpoint_frequency=100, change_feed=change_feed
  File "/mnt/commcare-hq-ro/corehq/apps/change_feed/consumer/feed.py", line 129, in __init__
    checkpoint_doc = self.checkpoint.get_or_create_wrapped()
  File "corehq/ex-submodules/pillowtop/checkpoints/manager.py", line 43, in get_or_create_wrapped
    checkpoint = get_or_create_checkpoint(self.checkpoint_id)
  File "corehq/ex-submodules/pillowtop/checkpoints/manager.py", line 23, in get_or_create_checkpoint
    'timestamp': datetime.utcnow()
```

**Comment by snopoke** [comment]
:white_check_mark: @emord @benrudolph made a few changes since last reivew

File: corehq/apps/change_feed/consumer/feed.py [hunk 6]
```diff
@@ -125,14 +126,6 @@ def __init__(self, checkpoint, checkpoint_frequency, change_feed):
         super(KafkaCheckpointEventHandler, self).__init__(checkpoint, checkpoint_frequency)
         assert isinstance(change_feed, KafkaChangeFeed)
         self.change_feed = change_feed
-        # todo: do this somewhere smarter?
```
Comment (benrudolph) [inline corehq/apps/change_feed/consumer/feed.py]: 👍

**Review by benrudolph — APPROVED** [review]

**Review by emord — APPROVED** [review]

**Review by benrudolph — APPROVED** [review]
looks good

**Comment by benrudolph** [comment]
(only review relevant changes to that file)

**Pull request merged**